vertical-pod-autoscaler/enhancements/7862-cpu-startup-boost/README.md

```diff
@@ -0,0 +1,283 @@
+
+## Proposal
+
+* To extend [`ContainerResourcePolicy`](https://github.com/kubernetes/autoscaler/blob/vertical-pod-autoscaler-1.3.0/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1/types.go#L191)
+with a new `StartupBoost` field to allow users to configure the CPU startup
+boost.
+
+* To extend [`ContainerScalingMode`](https://github.com/kubernetes/autoscaler/blob/vertical-pod-autoscaler-1.3.0/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1/types.go#L231-L236)
+with a new `StartupBoostOnly` mode to allow users to only enable the startup
+boost feature and not vanilla VPA altogether.
+
+## Design Details
+
+### Workflow
+
+1.  The user first configures the CPU startup boost on their VPA object
+
+1. When a pod targeted by that VPA is created, the kube-apiserver invokes the
+VPA Admission Controller
+
+1. The VPA Admission Controller modifies the pod's containers CPU request and
+limits to align with its `StartupBoost` policy, if specified, during the pod
+creation.
+
+1. The VPA Updater monitors pods targeted by the VPA object and when the pod
+condition is `Ready`, it scales down the CPU resources to the appropriate
+non-boosted value: `existing VPA recommendation for that container` (if any) OR
+the `CPU resources configured in the pod spec`.
+    * The scale down is applied [in-place](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources).
+
+### API Changes
+
+The new `StartupBoost` parameter will be added to the [`ContainerResourcePolicy`](https://github.com/kubernetes/autoscaler/blob/vertical-pod-autoscaler-1.3.0/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1/types.go#L191)
+and contain the following fields:
+  * `StartupBoost.CPU.Factor`: the factor by which to multiply the initial
+  resource request and limit of the containers' targeted by the VPA object.
+  * `StartupBoost.CPU.Value`: the target value of the CPU request or limit
+  during the startup boost phase.
+
+> [!IMPORTANT]
+> The boosted CPU value will be capped by
+> [`--container-recommendation-max-allowed-cpu`](https://github.com/kubernetes/autoscaler/blob/4d294562e505431d518a81e8833accc0ec99c9b8/vertical-pod-autoscaler/pkg/recommender/main.go#L122)
+> flag value, if set.
+
+> [!IMPORTANT]
+> Only one of `Factor` or `Value` may be specified per container policy.
+
+We will also add a new mode to the [`ContainerScalingMode`](https://github.com/kubernetes/autoscaler/blob/vertical-pod-autoscaler-1.3.0/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1/types.go#L231-L236):
+  * **NEW**: `StartupBoostOnly`: new mode that will allow users to only enable
+  the startup boost feature for a container and not vanilla VPA altogether.
+  * **NEW**: `Auto`: we will modify the existing `Auto` mode to enable both
```

Comment:
I left another comment similar to this:
Is `Auto` the one we want? At the moment we don't know when it will replace `InPlaceOrRecreate`

Reply:
Oh ... different auto. Ignore me!